### DISCUSSION TITLE
How to set permutations in TiledMMA for simt?

### DISCUSSION THREAD

**Discussion by cctry** [NVIDIA/cutlass#1362] [Q&A] [answered]
I am experimenting with the new TiledMMA interface and am confused about the permutation argument.

When letting 1 thread do 4x4x1 mma, I can set the tiler to be 4x4x1.
> tiled_mma = make_tiled_mma(UniversalFMA<float>>{}, Layout<Shape<_1,_1,_1>>{},  Tile<_4,_4,_1>{});
<img width="164" alt="image" src="https://github.com/NVIDIA/cutlass/assets/17473714/54ff8e4c-1b81-4a96-b6d1-a339664c9786">

Then, I replicate the atom to 2x2, so there should be 4 threads doing 8x8x1 mma. But the permutation is ignored here? 
> tiled_mma = make_tiled_mma(UniversalFMA<float>>{}, Layout<Shape<_1,_1,_1>>{},  Tile<_4,_4,_1>{});

<img width="164" alt="image" src="https://github.com/NVIDIA/cutlass/assets/17473714/642ff584-919f-42a9-b924-aec855dd42b6">

Expanding it to 16 threads also leads to an unexpected layout.
> tiled_mma = make_tiled_mma(UniversalFMA<float>>{}, Layout<Shape<_4,_4,_1>>{},  Tile<_4,_4,_1>{});
<img width="160" alt="image" src="https://github.com/NVIDIA/cutlass/assets/17473714/51feaa33-4736-4604-a3d9-67cc49607d25">

My understanding is that the size of each permutation is similar to the previous ValLayoutMNK, but it turns out to be different.

**Comment by thakkarV** [comment, marked as answer]
> Then, I replicate the atom to 2x2, so there should be 4 threads doing 8x8x1 mma. But the permutation is ignored here?

No, that is a misunderstanding. if you are creating a TiledMMA with an atom layout of 2x2 (seems like a copy paste error in your text) and a Tile of 4x4, each thread would compute only a 2x2 subtile. The tile describes the entire tile that this tiled MMA computes, not each thread. In general, you should not think of things in units of threads anyway, and rather in units of atoms.

**Comment by ccecka** [comment]
This thread has a good discussion that applies equally to SIMT Atoms:
https://github.com/NVIDIA/cutlass/discussions/1345
